### PR TITLE
feat: Allow implicit string → temporal conversion in SQL comparisons

### LINKED ISSUE #15956: Filtering dates in SQL raises a ComputeError
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from datetime import date
import polars as pl
import duckdb

df = pl.DataFrame({
    "order_date": [date(2023, 12, 1), date(2024, 1, 1)]
})

query = """
SELECT order_date
FROM df
WHERE order_date >= '2024-01-01'
"""

df.sql(query)
# ComputeError: cannot compare 'date/datetime/time' to a string value (create native python { 'date', 'datetime', 'time' } or compare to a temporal column)

duckdb.sql(query)
┌────────────┐
│ order_date │
│    date    │
├────────────┤
│ 2024-01-01 │
└────────────┘
```

### Log output

_No response_

### Issue description

Various SQL engines (PostgreSQL, BigQuery, duckdb) accept comparing a date with a string in the format 'YYYY-MM-DD'.

It would be nice if Polars also accepted this.

### Expected behavior

Return the filtered table
┌────────────┐
│ order_date │
│    date    │
├────────────┤
│ 2024-01-01 │
└────────────┘

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.23
Index type:           UInt32
Platform:             Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python:               3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.4
nest_asyncio:         1.5.7
numpy:                1.26.4
openpyxl:             3.1.2
pandas:               2.2.2
pyarrow:              15.0.1
pydantic:             2.4.2
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.23
xlsx2csv:             0.8.1
xlsxwriter:           3.1.8
```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #15956.

Where available, the internal active schema is now associated with `SQLExprVisitor`. This provides sufficient additional context for the binary comparison ops (and "IN" membership checks) to enable automatic temporal string casts (which PostgreSQL syntax allows). 

Handles comparison with bare column names (eg: `col > '2020-02-20'`) and cast expressions (eg: `expr::date > '2020-02-20'`), which catches most typical usage. We should be able to enhance further as needed.

#### Also:

Added `col::datetime` cast support (in addition to existing `col::timestamp`).

## Example

```python
import polars as pl

df = pl.DataFrame({
    "idx": [0, 1, 2],
    "dtm": [
        datetime(2024, 1, 7, 1, 2, 3, 123456),
        datetime(2006, 1, 1, 23, 59, 59, 555555),
        datetime(2020, 12, 30, 10, 30, 45, 987654),
    ],
    "dt": [
        date(2020, 12, 30),
        date(2077, 1, 1),
        date(1960, 1, 7),
    ],
})
```
```python
df.sql("SELECT * FROM self WHERE dtm < '2010-01-01 01:02:03'")
# shape: (1, 3)
# ┌─────┬────────────────────────────┬────────────┐
# │ idx ┆ dtm                        ┆ dt         │
# │ --- ┆ ---                        ┆ ---        │
# │ i64 ┆ datetime[μs]               ┆ date       │
# ╞═════╪════════════════════════════╪════════════╡
# │ 1   ┆ 2006-01-01 23:59:59.555555 ┆ 2077-01-01 │
# └─────┴────────────────────────────┴────────────┘
```
```python
df.sql("SELECT * FROM self WHERE dt IN ('1960-01-07','2077-01-01','2222-02-22')")
# shape: (2, 3)
# ┌─────┬────────────────────────────┬────────────┐
# │ idx ┆ dtm                        ┆ dt         │
# │ --- ┆ ---                        ┆ ---        │
# │ i64 ┆ datetime[μs]               ┆ date       │
# ╞═════╪════════════════════════════╪════════════╡
# │ 1   ┆ 2006-01-01 23:59:59.555555 ┆ 2077-01-01 │
# │ 2   ┆ 2020-12-30 10:30:45.987654 ┆ 1960-01-07 │
# └─────┴────────────────────────────┴────────────┘
```

### CODE REVIEW

**Comment by alexander-beedie** [comment]
Need to check/fix a few tests in the morning; parking in draft in the meantime…

**Comment by eitsupi** [comment]
I suppose this is not a specific to python? (`feat:` instead of `feat(python):`)

**Comment by alexander-beedie** [comment]
> I suppose this is not a specific to python? (`feat:` instead of `feat(python):`)

Quite right; brain clearly on autopilot there - fixed ;)

File: crates/polars-sql/Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,7 @@ polars-lazy = { workspace = true, features = ["abs", "binary_encoding", "concat_
 polars-plan = { workspace = true }
 
 hex = { workspace = true }
+once_cell = { workspace = true }
```
Comment (ritchie46) [inline crates/polars-sql/Cargo.toml:19]: We can use https://doc.rust-lang.org/std/sync/struct.OnceLock.html now. The `once_cell` crate has become redundant. We need to get that out.

File: crates/polars-sql/src/sql_expr.rs [hunk 2]
```diff
@@ -122,6 +129,7 @@ pub enum SubqueryRestriction {
 /// Recursively walks a SQL Expr to create a polars Expr
 pub(crate) struct SQLExprVisitor<'a> {
     ctx: &'a mut SQLContext,
+    active_schema: Option<SchemaRef>,
```
Comment (ritchie46) [inline crates/polars-sql/src/sql_expr.rs:132]: I think we can borrow here. `Option<&'a Schema>`.

File: crates/polars-sql/src/sql_expr.rs [hunk 3]
```diff
@@ -696,8 +755,25 @@ impl SQLExprVisitor<'_> {
...
 
+        let mut s = Series::from_any_values("", &list, true)?;
+
+        // handle implicit temporal strings, eg: "dt IN ('2024-04-30','2024-05-01')".
+        // (not yet as versatile as the temporal string conversions in visit_binary_op)
+        if s.dtype() == &DataType::String {
+            // handle implicit temporal string comparisons, eg: "dt >= '2024-04-30'"
+            if let Expr::Column(name) = &expr {
+                if self.active_schema.is_some() {
+                    let schema = self.active_schema.clone().unwrap();
```
Comment (ritchie46) [inline crates/polars-sql/src/sql_expr.rs:767]: Can borrow.

File: crates/polars-sql/src/sql_expr.rs [hunk 4]
```diff
@@ -345,17 +353,68 @@ impl SQLExprVisitor<'_> {
         }
     }
 
+    /// Handle implicit temporal string comparisons.
+    ///
+    /// eg: "dt >= '2024-04-30'", or "dtm::date = '2077-10-10'"
+    fn convert_temporal_strings(&mut self, left: &Expr, right: &Expr) -> Expr {
+        if let (Some(name), Some(s), expr_dtype) = match (&left, &right) {
```
Comment (ritchie46) [inline crates/polars-sql/src/sql_expr.rs:360]: `&left -> left` as we already borrow. Though I'd expect clippy to catch this one. Do we run it on this crate?

**Review by ritchie46 — COMMENTED** [review]
Nice one @alexander-beedie. I've left some comments.

File: crates/polars-sql/src/sql_expr.rs (hunk 4 above)
Comment (alexander-beedie) [reply crates/polars-sql/src/sql_expr.rs]: Yup; it caught a few other minor things in this PR, heh

File: crates/polars-sql/src/sql_expr.rs (hunk 2 above)
Comment (alexander-beedie) [reply crates/polars-sql/src/sql_expr.rs]: Done (I think :))

File: crates/polars-sql/Cargo.toml (hunk 1 above)
Comment (alexander-beedie) [reply crates/polars-sql/Cargo.toml:19]: Looks like we're using `once_cell::sync::Lazy` in quite a few places (17 according to a quick grep through the code) - if the replacement pattern is `LazyLock` that hasn't been stabilised yet 🤔

**Comment by alexander-beedie** [comment]
FYI: the failures in `test_read_kuzu_graph_database` are fixed by https://github.com/pola-rs/polars/pull/16031.

File: crates/polars-sql/Cargo.toml (hunk 1 above)
Comment (ritchie46) [reply crates/polars-sql/Cargo.toml:19]: No, this is what I meant. https://doc.rust-lang.org/std/sync/struct.OnceLock.html

We use that on a few places already. You create a `OnceLock` and a function `get_locked` which gets or intits.

Comment (alexander-beedie) [reply crates/polars-sql/Cargo.toml:19]: Ahh, nice; will update later this evening 👌

Comment (alexander-beedie) [reply crates/polars-sql/Cargo.toml:19]: Done 👌

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**